### PR TITLE
Trial run of adding newbase back

### PR DESCRIPTION
**Pull request by kshyatt**
Addresses part of #19839

### CODE REVIEW

File: base/libgit2/libgit2.jl [hunk 1]
```diff
@@ -489,10 +489,15 @@ function rebase!(repo::GitRepo, upstream::AbstractString="")
         else
             GitAnnotated(repo, upstream)
         end
+        onto_ann  = if isempty(newbase)
+                        Nullable{GitAnnotated}()
+                    else
+                        Nullable{GitAnnotated}(GitAnnotated(repo, newbase))
+        end
```
Comment (ararslan) [inline base/libgit2/libgit2.jl]: This seems stylistically different than what's around elsewhere in Base (or maybe it's just that the `else` and `end` aren't in the same column that's bugging me). Perhaps something like
```julia
onto_ann = if isempty(newbase)
    Nullable{GitAnnotated}()
else
    Nullable{GitAnnotated}(GitAnnotated(repo, newbase))
end
```
would be acceptable? Not really a big deal.

Comment (dmbates) [inline base/libgit2/libgit2.jl]: or
```jl
onto_ann = isempty(newbase) ? Nullable{GitAnnotated}() :
    Nullable{GitAnnotated}(GitAnnotated(repo, newbase))
```

File: base/libgit2/libgit2.jl [hunk 2]
```diff
@@ -457,7 +457,7 @@ function merge!(repo::GitRepo;
 end
 
 """
-    LibGit2.rebase!(repo::GitRepo[, upstream::AbstractString])
+    LibGit2.rebase!(repo::GitRepo, upstream::AbstractString="", newbase::AbstractString="")
 
 Attempt an automatic merge rebase of the current branch, from `upstream` if provided, or
 otherwise from the upstream tracking branch.
```
Comment (simonbyrne) [inline base/libgit2/libgit2.jl]: describe what `newbase` does here.

File: base/libgit2/libgit2.jl (hunk 1 above)
Comment (simonbyrne) [inline base/libgit2/libgit2.jl]: Or even change the `onto` argument in `GitRebase` to accept a `Union{Void, GitAnnotated}`? I don't think the cost of dynamic dispatch is going to be to burdensome.

**Review by simonbyrne — CHANGES_REQUESTED** [review]
Maybe add a test as well?

**Comment by kshyatt** [comment]
I'm a bit confused about how to test this, honestly.

**Comment by simonbyrne** [comment]
Ideally, I guess, we could compare the results to command line git.
(to be honest though, I'm still not 100% sure what exactly this option is supposed to do)

**Comment by tkelman** [comment]
what does this correspond to in the libgit2 api? if they don't have it documented we can bug them about it

**Comment by kshyatt** [comment]
Test added!

File: base/libgit2/libgit2.jl (hunk 1 above)
Comment (kshyatt) [inline base/libgit2/libgit2.jl]: Better now?

**Comment by ararslan** [comment]
> I don't know if we want ASCII art in the docstring?

Since these functions aren't exported from Base, I think they're more for people who know what they're doing. As such, I wouldn't think we need that level of detail in the docstrings. If anything, we could simply link to the relevant page in the Git documentation.

File: base/libgit2/libgit2.jl [hunk 3]
```diff
@@ -457,10 +457,11 @@ function merge!(repo::GitRepo;
...
+`newbase` is the branch to rebase onto, or `NULL` to rebase onto the given `upstream`.
```
Comment (simonbyrne) [inline base/libgit2/libgit2.jl]: maybe:
```
`newbase` is the branch to rebase onto, by default this is `upstream`.
```

**Comment by simonbyrne** [comment]
LGTM other than the doc wording.

**Review by simonbyrne — APPROVED** [review]

**Pull request merged**